aiida/manage/manager.py

```diff
@@ -123,8 +124,10 @@ def get_backend_manager(self):
         :return: the database backend manager
         :rtype: :class:`aiida.backend.manager.BackendManager`
         """
+        from aiida.backends import get_backend_manager
+
         if self._backend_manager is None:
-            from aiida.backends import get_backend_manager
+            self._load_backend()
```

Comment:
`get_backend_manager` should load the backend?

isn't that a significant change?
should this be in this PR?

I probably just miss the reasoning for these changes, could you please explain?

Reply:
All the changes in this file can be explained as follows, and they should answer all three of your questions. The `get_backend_manager` was so far _not_ loading the backend, although that really doesn't make any sense. It is providing access to data from the database, but to do so the backend should be loaded, otherwise a connection isn't possible. This problem went unnoticed, because the `BackendManager` is only used in `aiida.engine.utils.set_process_state_change_timestamp`. By the time this gets used, the database backend will already have been loaded through another code path.

For the changes in this PR, however, I needed to get _just_ the backend manager and have the backend loaded, which wasn't the case. I couldn't simply call `_load_backend()` in the `get_backend_manager` because this would lead to infinite recursion as `_load_backend()` also calls `get_backend_manager`. Therefore I had to refactor `_load_backend` to not call the former and directly fetch it through `aiida.backends.

Reply:
Thanks a lot for the clarification, very helpful!

One last question: `_load_backend()` now uses a backend manager that is then thrown away.
If these two things indeed have to go together, why not let `_load_backend()` also set `self._backend_manager`?

The current solution seems to duplicate the creation of the backend manager, once in `_load_backend()` and once in `get_backend_manager()`.

Reply:
I left it in because I wasn't sure that there wasn't a code path (unlikely, but still) where the backend manager instance would _not_ be constructed in `_load_backend` because it is in the conditional `if configuration.BACKEND_UUID is None:`. If that somehow is `False` the calling `_load_backend` will not create an instance of the backend manager. So calling `get_backend_manager` will return `None`. I have looked more into the code to see where the design mistake is, but I am not sure yet how to improve the flow. Most of this code that makes it complicated is to deal with the possibility of unloading/reloading different profiles, which is currently anyway not supported. Maybe it is best to leave like this for now, but if you prefer I can assign the backend manager in `_load_backend` to `self._backend_manager` and remove the construction from `get_backend_manager`.

Reply:
Ok, please add a comment in the corresponding issue so that we don't forget

Reply:
[Done!](https://github.com/aiidateam/aiida-core/issues/3011)